Candidate enhancements to log_file_checks

bieryAtFnal opened the pull request:
Last week, during the testing of the changes that we made to split out the DQM process from the Readout process, I noticed some changes in the log_file_checks function that I would have found helpful.  This PR has candidate implementations for those changes.

1. The existing logic in the `logs_are_error_free` function (`all_ok=all_ok and log_has_no_errors(log)`) would stop checking for errors after the first logfile in which there was an error.  I would have found it helpful for the test to check *all* logfiles, so I've added a flag (`show_all_problems`) to enable that behavior.
2. I would find it helpful to know which logfile contained which errors, so I added a flag to enable the printout of the logfile name when there is one or more errors in a logfile.
3. It would be *very* helpful to be able to tell the `log_file_check` function to ignore certain 'expected' problems.  This is implemented in the  new `excluded_substring_map` argument.

eflumerf commented on python/integrationtest/log_file_checks.py:
```diff
@@ -1,31 +1,74 @@
+
+        if bad_line:
+            ignore_this_problem=False
+            for excluded_substring in excluded_substring_list:
```
Is this the sort of thing we may want the power of regexes for? I think exclusion strings should be as specific as possible, and regular expressions allow for skipping parts of the message which may change (counters, etc), while also allowing certain iterations of the message to go through (exclude if counter is below 100, but not if its over).

eflumerf reviewed (approved):
LGTM

philiprodrigues commented on python/integrationtest/log_file_checks.py:
```diff
@@ -1,31 +1,74 @@
+#   dictionary keyed by strings that might appear in the logfile name and having values
+#   that are lists of excluded phrases. For example:
+#   ex_sub_map = {"ruemu": ["expected problem phrase 1", "expected problem  phrase 2"]}
+def logs_are_error_free(log_file_names, show_all_problems=False, print_logfilename_for_problems=False,
```
Do you want to make `show_all_problems` and `print_logfilename_for_problems` both default to `True`? I think that would be reasonable

philiprodrigues reviewed (approved):
I think this all looks sensible. I agree with Eric that regexes for the excluded phrases would be ideal, but I'm OK with merging this now and leaving that for another time if you prefer.

bieryAtFnal commented:
Thanks, guys, for your feedback.

The regex idea sounds quite sensible to me, but I want to think about that a little bit.  I've found that getting the correctly-escaped string for Python regex is not always trivial, so I'd want to provide some examples for users, and I'll need to work out those examples.

As for whether to make the defaults for the boolean arguments True, that sounds fine to me.  (Actually, I wondered if you might say that we should simply change the behavior of the code and skip the flags :) ).  
If we make the defaults for `show_all_problems` and `print_logfilename_for_problems` True, I may change the order of the arguments to the `logs_are_error_free` function so that they are last...

bieryAtFnal commented:
I have added support for regular expressions, and I have switched the defaults for the "print" arguments to True.
I believe that the changes are now ready to be merged to develop.  Phil, if you would take another look at the changes and do the merge whenever it is convenient, that would be great.
In addition, I wonder how we would then make these changes available in the nightly builds.  Would we just ask Pengfei to make that happen, or does this package need to be tagged?
Thanks again,
Kurt

philiprodrigues commented:
Sorry for the delay in testing. This all looks good to me. I tested with four_process_quick-test by adding a dummy warning message to TriggerRecordBuilder. That initially _didn't_ fail the integration test, because the code was looking for the severity to be printed as "WARN" instead of "WARNING", so I took the liberty of fixing that.

With that fix, we fail the test. I can make it pass again by adding `excluded_substring_map={"dataflow": [".*Dummy.*"]})` to the `logs_are_error_free` call.
